src/mono/mono/mini/aot-compiler.c

```diff
@@ -13215,6 +13220,124 @@ add_profile_instances (MonoAotCompile *acfg, ProfileData *data)
 	printf ("Added %d methods from profile.\n", count);
 }
 
+static int
+add_single_mibc_profile_method (MonoAotCompile *acfg, MonoMethod *method)
+{
+	if (!method->is_inflated || mono_method_is_generic_sharable_full (method, FALSE, FALSE, FALSE)) {
+		if (!acfg->aot_opts.profile_only)
+			return 0;
+		if (m_class_get_image (method->klass) != acfg->image)
+			return 0;
+		add_profile_method (acfg, method);
+		return 1;
+	}
+
+	if (acfg->aot_opts.dedup_include) {
+		add_profile_method (acfg, method);
+		return 1;
+	}
+
+	MonoGenericContext *ctx = mono_method_get_context (method);
```

Comment:
Can we make a common function for this logic that could be shared with the legacy profiler?

Also, let's not lose the comment about what it is doing - if we ever delete the legacy profiler, this code won't have an explanation in the source anymore.

Reply:
Made this function `add_single_profile_method` the common function that is used for both legacy and new profiler.

The legacy profiler would loop through all methods twice, call `add_profile_method (acfg, m)` when conditions are satisfactory (filtering out certain methods), and maintaining a count of how many methods are successfully added.

The new profiler obtains each profile method one at a time, so instead, the filter for whether or not a method is deemed satisfactory is extracted into this function `add_single_profile_method`.

Legacy
```
if (acfg->aot_opts.profile_only) {
        /* Add methods referenced by the profile */
        g_hash_table_iter_init (&iter, data->methods);
        while (g_hash_table_iter_next (&iter, &key, &value)) {
                MethodProfileData *mdata = (MethodProfileData*)value;
                MonoMethod *m = mdata->method;

                if (!m)
                        continue;
                if (m->is_inflated)
                        continue;
                if (m_class_get_image (m->klass) != acfg->image)
                        continue;
                add_profile_method (acfg, m);
                count ++;
        }
}

/*
* Add method instances 'related' to this assembly to the AOT image.
*/
g_hash_table_iter_init (&iter, data->methods);
while (g_hash_table_iter_next (&iter, &key, &value)) {
        MethodProfileData *mdata = (MethodProfileData*)value;
        MonoMethod *m = mdata->method;
        MonoGenericContext *ctx;

        if (!m)
                continue;
        if (!m->is_inflated)
                continue;
        if (mono_method_is_generic_sharable_full (m, FALSE, FALSE, FALSE)) {
                if (acfg->aot_opts.profile_only && m_class_get_image (m->klass) == acfg->image) {
                        // Add the fully shared version to its home image
                        add_profile_method (acfg, m);
                        count ++;
                }
                continue;
        }

        if (acfg->aot_opts.dedup_include) {
                /* Add all instances from the profile */
                add_profile_method (acfg, m);
                count ++;
        } else {
                ctx = mono_method_get_context (m);
                /* For simplicity, add instances which reference the assembly we are compiling */
                if (((ctx->class_inst && inst_references_image (ctx->class_inst, acfg->image)) ||
                                (ctx->method_inst && inst_references_image (ctx->method_inst, acfg->image)))) {
                        //printf ("%s\n", mono_method_full_name (m, TRUE));
                        add_profile_method (acfg, m);
                        count ++;
                } else if (m_class_get_image (m->klass) == acfg->image &&
                                        ((ctx->class_inst && is_local_inst (ctx->class_inst, acfg->image)) ||
                                        (ctx->method_inst && is_local_inst (ctx->method_inst, acfg->image)))) {
                        /* Add instances where the gtd is in the assembly and its inflated with types from this assembly or corlib */
                        //printf ("%s\n", mono_method_full_name (m, TRUE));
                        add_profile_method (acfg, m);
                        count ++;
                } else {
                        //printf ("SKIP: %s (%s)\n", mono_method_get_full_name (m), acfg->image->name);
                }
                /*
                 * FIXME: We might skip some instances, for example:
                 * Foo<Bar> won't be compiled when compiling Foo's assembly since it doesn't match the first case,
                 * and it won't be compiled when compiling Bar's assembly if Foo's assembly is not loaded.
                 */
        }
}
```

The easiest differentiator between the first block of `methods referenced in the profile` and second block of `methods related to this assembly` is `method->is_inflated`, where the former doesn't `continue` if `is_inflated` is true, and the latter doesn't `continue` if `is_inflated` is false. 

With that differentiation, both the `/* Add methods referenced by the profile */` block and `// Add the fully shared version to its home image` require `acfg->aot_opts.profile_only` and `m_class_get_image (method->klass) == acfg->image` to be true in order to add the method, so theyre combined and we early return if either condition are false.

Finally, the latter bulk of `Add method instances 'related' to this assembly to the AOT image.` is kept the same, except the `else if` and `else` are removed in favor of early returns.